### DISCUSSION TITLE
WebpackError: ReferenceError: self is not defined

### DISCUSSION THREAD

**Discussion by ashishch-98** [gatsbyjs/gatsby#27741] [Help]
I am using Firebase Cloud Messaging with GatsbyJs. For setting up FCM I have used [gatsby-plugin-firebase-messaging](https://www.gatsbyjs.com/plugins/gatsby-plugin-firebase-messaging/#:~:text=gatsby-plugin-firebase-messaging%20This%20plugin%20handles%20the%20setup%2C%20configuration%2C%20and%2C,can%20be%20done%20as%20outlined%20in%20Firebase%E2%80%99s%20documentation.). So on `gatsby develop` everything works as it should, but on `gatsby-build` it starts to break. 

The error on build was _**WebpackError: ReferenceError: IDBIndex is not defined**_.  I used the solution given in this [thread](https://github.com/firebase/firebase-js-sdk/issues/2222#issuecomment-538072948) and the error was eliminated but now a new error is there during the build. 

```
See our docs page for more info on this error: https://gatsby.dev/debug-html


  190 |     }
  191 | 
> 192 |     return React.createElement.apply(null, createElementArgArray);
      | ^
  193 |   }
  194 | 
  195 |   return React.createElement.apply(null, args);


  WebpackError: ReferenceError: self is not defined

  - esm.js:192 
    node_modules/@mdx-js/react/dist/esm.js:192:1

  - PushNotifications.js:26 
    src/components/PushNotifications.js:26:34

```

gatsby-node.js
```
exports.onCreateWebpackConfig = ({
  stage,
  actions,
  getConfig
}) => {
  if (stage === 'build-html') {
    actions.setWebpackConfig({
      externals: getConfig().externals.concat(function(context, request, callback) {
        const regex = /^@?firebase(\/(.+))?/;
        // exclude firebase products from being bundled, so they will be loaded using require() at runtime.
        if (regex.test(request)) {
          return callback(null, 'umd ' + request);
        }
        callback();
      })
    });
  }
};
```

Why it is there and how can it be solved?

**Comment by nibtime** [comment]
@ashishch-98 

You can prevent such problems altogether with dynamic imports of Firebase services and without using a Gatsby plugin. I provided [an answer in another discussion](https://github.com/gatsbyjs/gatsby/discussions/27957#discussioncomment-397948) that shows how to do that.
